source/Nevermore/IQueryBuilder.cs

```diff
@@ -1,40 +1,85 @@
 using System;
 using System.Collections.Generic;
-using Nevermore.Joins;
+using Nevermore.AST;
 
 namespace Nevermore
 {
-    public interface IQueryBuilder<TRecord> where TRecord : class
+    public interface IOrderedQueryBuilder<TRecord> : IQueryBuilder<TRecord> where TRecord : class
+    {
+    }
+
+    public interface ITableSourceQueryBuilder<TRecord> : IQueryBuilder<TRecord> where TRecord : class
     {
-        IQueryGenerator QueryGenerator { get; }
+        ITableSourceQueryBuilder<TRecord> View(string viewName);
+        ITableSourceQueryBuilder<TRecord> Table(string tableName);
+        ITableSourceQueryBuilder<TRecord> Alias(string tableAlias); 
+        IQueryBuilder<TRecord> Hint(string tableHint);
+        IAliasedSelectSource AsAliasedSource();
+    }
 
+    public interface ISubquerySourceBuilder<TRecord> : IQueryBuilder<TRecord> where TRecord : class
+    {
+        ISubquerySource AsSource();
+        ISubquerySourceBuilder<TRecord> Alias(string subqueryAlias);
+    }
+
+    public interface IJoinSourceQueryBuilder<TRecord> : IQueryBuilder<TRecord> where TRecord : class
+    {
+        IJoinSourceQueryBuilder<TRecord> On(string leftField, JoinOperand operand, string rightField);
+    }
+
+    public interface IQueryBuilder<TRecord> where TRecord : class
+    {
         IQueryBuilder<TRecord> Where(string whereClause);
-        IQueryBuilder<TRecord> Where(string fieldName, SqlOperand operand, object value);
-        IQueryBuilder<TRecord> Where(string fieldName, SqlOperand operand, object startValue, object endValue);
-        IQueryBuilder<TRecord> Where(string fieldName, SqlOperand operand, IEnumerable<object> values);
-        
+        IQueryBuilder<TRecord> WhereParameterised(string fieldName, UnarySqlOperand operand, Parameter parameter);
+        IQueryBuilder<TRecord> WhereParameterised(string fieldName, BinarySqlOperand operand, Parameter startValueParameter, Parameter endValueParameter);
+        IQueryBuilder<TRecord> WhereParameterised(string fieldName, ArraySqlOperand operand, IEnumerable<Parameter> parameterNames);
+
         IOrderedQueryBuilder<TRecord> OrderBy(string orderByClause);
         IOrderedQueryBuilder<TRecord> OrderByDescending(string orderByClause);
+        IQueryBuilder<TRecord> IgnoreDefaultOrderBy();
```

Comment:
Is this needed? Previously we had a default orderby (`ID`) that was used if the user didn't specify an `OrderBy`. This method implies the new behaviour that it always adds `ID` to the order by unless you call this method.

I can't think of a situation where you want to not have an order by

Reply:
I believe that some of our stored procs do not have a default order by. For example, the dashboard query.

Reply:
They are rather the exception as the Sprocs/views do ordering internally

Reply:
Also if there is no ordering currently, I see no harm in ordering by `ID`

Reply:
Fair enough, I'll remove this :+1:

Reply:
Done